### PR TITLE
[inductor] Low memory dropout

### PR DESCRIPTION
**Pull request by jansel**
Forwards (from test_lowmem_dropout_cuda)
```
@pointwise_heuristics(size_hints=[256])
@triton.jit
def kernel0(seed0, in_ptr1, out_ptr0, ks0, ks1, xnumel, XBLOCK : tl.constexpr):
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.reshape(tl.arange(0, XBLOCK), [XBLOCK])
    xmask = xindex < xnumel
    x0 = xindex
    tmp0 = tl.load(seed0 + 0, None)
    tmp6 = tl.load(in_ptr1 + x0, xmask)
    tmp1 = x0
    tmp2 = tl.rand(tmp0, tmp1)
    tmp3 = 0.5
    tmp4 = tmp2 > tmp3
    tmp5 = tmp4.to(tl.float32)
    tmp7 = tmp5 * tmp6
    tmp8 = 2.0
    tmp9 = tmp7 * tmp8
    tl.store(out_ptr0 + x0, tmp9, xmask)

@pointwise_heuristics(size_hints=[256])
@triton.jit
def kernel1(seed0, in_ptr1, out_ptr0, ks0, ks1, xnumel, XBLOCK : tl.constexpr):
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.reshape(tl.arange(0, XBLOCK), [XBLOCK])
    xmask = xindex < xnumel
    x0 = xindex
    tmp0 = tl.load(seed0 + 0, None)
    tmp6 = tl.load(in_ptr1 + x0, xmask)
    tmp1 = 256 + x0
    tmp2 = tl.rand(tmp0, tmp1)
    tmp3 = 0.5
    tmp4 = tmp2 > tmp3
    tmp5 = tmp4.to(tl.float32)
    tmp7 = tmp5 * tmp6
    tmp8 = 2.0
    tmp9 = tmp7 * tmp8
    tl.store(out_ptr0 + x0, tmp9, xmask)


def call(primals_1, primals_2, primals_3):
    torch.randint(2**31, size=(), dtype=torch.int32, out=seed_cuda_0)
    primals_1_size = primals_1.size()
    s0 = primals_1_size[0]
    primals_3_size = primals_3.size()
    s1 = primals_3_size[0]
    buf0 = empty_strided((s1, s0), (s0, 1), device='cuda', dtype=torch.float32)
    aten.mm.out(primals_3, as_strided(primals_1, (s0, s0), (1, s0)), out=buf0)
    buf1 = empty_strided((s1, s0), (s0, 1), device='cuda', dtype=torch.float32)
    kernel0_xnumel = s0*s1
    kernel0[grid(kernel0_xnumel)](seed_cuda_0, buf0, buf1, s0, s1, kernel0_xnumel)
    buf2 = buf0; del buf0
    aten.mm.out(buf1, as_strided(primals_2, (s0, s0), (1, s0)), out=buf2)
    buf3 = empty_strided((s1, s0), (s0, 1), device='cuda', dtype=torch.float32)
    kernel1_xnumel = s0*s1
    kernel1[grid(kernel1_xnumel)](seed_cuda_0, buf2, buf3, s0, s1, kernel1_xnumel)
    return (buf3, primals_3, seed_cuda_0.clone(), buf1, as_strided(primals_2, (s0, s0), (s0, 1)), )
```

Backwards:
```
@pointwise_heuristics(size_hints=[256])
@triton.jit
def kernel0(in_ptr0, in_ptr1, out_ptr0, ks0, ks1, xnumel, XBLOCK : tl.constexpr):
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.reshape(tl.arange(0, XBLOCK), [XBLOCK])
    xmask = xindex < xnumel
    x0 = xindex
    tmp0 = tl.load(in_ptr0 + 0, None)
    tmp6 = tl.load(in_ptr1 + x0, xmask)
    tmp1 = 256 + x0
    tmp2 = tl.rand(tmp0, tmp1)
    tmp3 = 0.5
    tmp4 = tmp2 > tmp3
    tmp5 = tmp4.to(tl.float32)
    tmp7 = tmp5 * tmp6
    tmp8 = 2.0
    tmp9 = tmp7 * tmp8
    tl.store(out_ptr0 + x0, tmp9, xmask)

@pointwise_heuristics(size_hints=[256])
@triton.jit
def kernel1(in_ptr0, in_ptr1, out_ptr0, ks0, ks1, xnumel, XBLOCK : tl.constexpr):
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.reshape(tl.arange(0, XBLOCK), [XBLOCK])
    xmask = xindex < xnumel
    x0 = xindex
    tmp0 = tl.load(in_ptr0 + 0, None)
    tmp6 = tl.load(in_ptr1 + x0, xmask)
    tmp1 = x0
    tmp2 = tl.rand(tmp0, tmp1)
    tmp3 = 0.5
    tmp4 = tmp2 > tmp3
    tmp5 = tmp4.to(tl.float32)
    tmp7 = tmp5 * tmp6
    tmp8 = 2.0
    tmp9 = tmp7 * tmp8
    tl.store(out_ptr0 + x0, tmp9, xmask)


def call(primals_3, philox_seed_like_default, mul_tensor_1, permute_default_4, tangents_1):
    primals_3_size = primals_3.size()
    s0 = primals_3_size[0]
    s1 = primals_3_size[1]
    buf0 = empty_strided((s0, s1), (s1, 1), device='cuda', dtype=torch.float32)
    kernel0_xnumel = s0*s1
    kernel0[grid(kernel0_xnumel)](philox_seed_like_default, tangents_1, buf0, s0, s1, kernel0_xnumel)
    buf1 = empty_strided((s1, s1), (s1, 1), device='cuda', dtype=torch.float32)
    aten.mm.out(as_strided(buf0, (s1, s0), (1, s1)), mul_tensor_1, out=buf1)
    buf2 = empty_strided((s0, s1), (s1, 1), device='cuda', dtype=torch.float32)
    aten.mm.out(buf0, permute_default_4, out=buf2)
    buf3 = as_strided(buf0, (s1, s0), (1, s1)); del buf0
    kernel1_xnumel = s0*s1
    kernel1[grid(kernel1_xnumel)](philox_seed_like_default, buf2, buf3, s0, s1, kernel1_xnumel)
    del buf2
    buf4 = empty_strided((s1, s1), (s1, 1), device='cuda', dtype=torch.float32)
    aten.mm.out(buf3, primals_3, out=buf4)
    return (as_strided(buf4, (s1, s1), (s1, 1)), as_strided(buf1, (s1, s1), (s1, 1)), None, )

```

### CODE REVIEW

File: torchinductor/overrides.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+import torch
+from torch import _prims
+
+log = logging.getLogger(__name__)
+
+
+@contextlib.contextmanager
+def patch_functions():
+    prior = torch.nn.functional.dropout
+    torch.nn.functional.dropout = lowmem_dropout
```
Comment (jansel) [inline torchinductor/overrides.py:15]: @Chillee is there a better way to do decomposition *before* autograd?

For this I want to change the backwards formula to pass different things.  For now, I'm doing it at the python level, but I feel like this is something AOT Autograd should be able to do for me in a more robust way.

Comment (jansel) [reply torchinductor/overrides.py:15]: If we had mechanism for swapping out functions before grad we could also auto-generate backwards from the forwards decomp.

Comment (Chillee) [reply torchinductor/overrides.py:15]: Hmm, not super easily. Maybe @ezyang has any thoughts?

In the past I considered tracing out the forwards pass first so we could transform the forwards pass first before transforming the backwards pass.

I also remember @ngimel had some concerns about doing this kind of dropout vs. just saving the seed directly.

I also wonder whether we should just do this kind of transformation in the graph. For example, if we were to rewrite the `randn_like` in the joint fw/bw graph to take a seed, and then allowed rematerializing `randn_like`, the partitioning algorithm would naturally work for this.

Comment (ngimel) [reply torchinductor/overrides.py:15]: Yeah with our typical approach (and the one nvfuser is using) making sure that with cuda graphs the rng will get into correct state is hard. In inductor we are setting the state via a separate call to `randint` (launches extra kernel) anyway, so it's entirely possible (but we are launching an extra kernel always, which the regular approach avoids)

Comment (ezyang) [reply torchinductor/overrides.py:15]: You don't have to monkeypatch the function, just define a TorchFunctionMode. But yes interposing at the Python level is the easiest way to get in before autograd. We have discussed more flexibility in interposing on dispatch but nothing implemented yet.

Comment (jansel) [reply torchinductor/overrides.py:15]: @ezyang do you have an example of TorchFunctionMode usage?

Comment (ezyang) [reply torchinductor/overrides.py:15]: https://github.com/albanD/subclass_zoo/blob/main/autograd_monkeypatch.py and yes it will work with AOT Autograd

**Review by ngimel — APPROVED** [review]
